lib/gssapi/krb5/acquire_cred.c

```diff
@@ -80,6 +80,34 @@ get_keytab(krb5_context context, krb5_keytab *keytab)
     return (kret);
 }
 
+static krb5_error_code
+get_client_keytab(krb5_context context,
+		  krb5_const_principal principal,
+		  krb5_keytab *keytab)
+{
+    krb5_error_code kret;
+    krb5_keytab_entry entry;
+    char *name = NULL;
+
+    kret = _krb5_kt_client_default_name(context, &name);
+    if (kret != 0)
+	return kret;
+
+    kret = krb5_kt_resolve(context, name, keytab);
+    if (kret == 0 && principal != NULL) {
+	kret = krb5_kt_get_entry(context, *keytab, principal,
+				 0, 0, &entry);
+	if (kret == 0)
+	    krb5_kt_free_entry(context, &entry);
+    }
+    krb5_xfree(name);
+
+    if (kret != 0)
+	kret = get_keytab(context, keytab);
```

Comment:
If a principal name was given, do we not want to check if this keytab has entries for it?

Reply:
Probably but – didn't want to change current behavior. Hmm, does Heimdal do the MIT thing where rd_req tries all the keys in the keytab? We'd have to be careful we don't break that (if it does that).

Reply:
Don't think it'll make much difference though for the client, it'd just fail later on.

Reply:
Hmm I guess we don’t handle the case where we don’t know the client principal name. Todo.